### ISSUE TITLE
Message not displayed when tapping notification in iOS 10

### ISSUE THREAD

**Issue by phatmann** [intercom/intercom-ios#184]
Lots of folks are having issues with push notifications on iOS 10, because Apple broke existing behavior. The app delegate method `application:didReceiveRemoteNotification:userInfo:fetchCompletionHandler` is not called when the app is in the background. We have a workaround in our app for this, but unfortunately this is the method you guys "swizzled" in your SDK to intercept push notifications. Any chance of a fix? Apple fixed this in iOS 10.1 but we don't want to wait for that.

## Version info
  - Intercom for iOS version: 3.0.12
  - iOS version: 10.0

## Expected behavior

When app is in background, tapping a notification from Intercom should display chat message in the app.

## Actual behavior

App does not display chat message.

## Steps to reproduce
 1. Background app.
 2. Send push from Intercom.
 3. Tap notification

**Comment by phatmann** [comment]
One fix: swizzle the new UserNotificationCenter methods.

**Comment by jtreanor** [comment]
Thanks for the report @phatmann,

We're aware of these issues. I will dive into this tomorrow and keep you updated on a fix.

**Comment by jtreanor** [comment]
@phatmann I'm currently working on putting together a solution to this that will have the minimal impact on users.

Do you implement `UNUserNotificationCenterDelegate` and in your app?
If you do could you temporarily remove it and see if Intercom push starts to work?

I'm quite confident I have tracked down the scenario and fix for this but it would be great if you could confirm this.

**Comment by phatmann** [comment]
I did implement UNUserNotificationCenterDelegate. I will try it without that and see what happens.

**Comment by phatmann** [comment]
The behavior is the same with or without using UNUserNotificationCenterDelegate. The app comes to the foreground, but it does not launch into the messenger window. However, as of today, the message bubble appears at bottom, when yesterday it did not. Did you fix something on your back end?

**Comment by jtreanor** [comment]
Thanks for checking @phatmann.

That is interesting and not quite what I have seen in my testing. Unfortunately we are somewhat in the dark on how Apple has changed this behaviour.

I have prepared a fix that also intercepts push from `UNUserNotificationCenterDelegate` which works 100% in my testing but I want to be fully sure before we go ahead and release it.

Some more questions for you if you have a second:

1. Which of these methods have you implemented in your app delegate?
    - `application:didReceiveRemoteNotification:`
    - `application:didReceiveRemoteNotification:fetchCompletionHandler:`
2. Does your app have `remote-notifications` in the `UIBackgroundModes` section of your `Info.plist`?
3. Is the Push Notifications entitlement in your apps capabilities file (check in Targets->Capabilities)?
4. Do you use any other third party push notifications provider?

**Comment by phatmann** [comment]
I implemented `application:didReceiveRemoteNotification:fetchCompletionHandler:` to handle notifications on iOS 9. I did not implement `application:didReceiveRemoteNotification:` since it is deprecated. Ironically, `application:didReceiveRemoteNotification:` is still called even when running in iOS 10, but I did not exploit that oddity.

My app receives push notifications via AWS SNS just fine. It has `remote-notifications` background mode and has the push notifications entitlement. The problem is just for Intercom notifications.

Hope this helps. Really glad you have a fix cooked up. Apple really did us in on this one 🙃

**Comment by jtreanor** [comment]
Thanks for your help @phatmann.

I'll keep you updated on the fix by posting here 😄

**Comment by jtreanor** [comment]
The fix I described has been released in [3.0.14](https://github.com/intercom/intercom-ios/releases/tag/3.0.14).

This fixes iOS 10 push compatibility for every example case I can come up with. However, I'll wait until you verify it fixes issue for you before closing it out. I'd appreciate if you could test both with & without `UNUserNotificationCenterDelegate`.

Thanks again for all your help @phatmann.

**Comment by phatmann** [comment]
I will try it out now. What is the correct behavior, for the messenger window to be displayed?

**Comment by jtreanor** [comment]
@phatmann Yes. Thats the expected behaviour.

**Comment by phatmann** [comment]
I just tested this out. The behavior, at least for me, is unchanged with the new 3.0.14 library. The bubble appears, but the messenger window does not.

My app delegate code, (since this is clearly the next thing you will ask me for 😁)

```swift
    open func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        if #available(iOS 10.0, *) {
            UNUserNotificationCenter.current().delegate = self
        }
        
        return true
    }

    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Swift.Void) {
        completionHandler([])
        NotificationHandler.notificationReceived(notification.request.content.userInfo)
    }
    
    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Swift.Void) {
        completionHandler()
        NotificationHandler.notificationTapped(response.notification.request.content.userInfo)
    }
```

**Comment by jtreanor** [comment]
😞

Thanks for the detail, I know we can track this down 😄 

The code looks perfect, there are no obvious issues with it. For completeness, does your class definition include `UNUserNotificationCenterDelegate` something like this?

```
class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate {
    ...
}
```

I will replicate your app delegate as close as possible and get back to you.

**Comment by phatmann** [comment]
Yes, my AppDelegate does implement the `UNUserNotificationCenterDelegate` protocol.

A few more data points:

* When the app is in the foreground, the  `userNotificationCenter:willPresent: withCompletionHandler` method is not being intercepted by the Intercom library. This means our foreground notification UI gets called. When the "Show" button in our dialog is tapped, the messenger bubble appears. This is a bug we can live with, but it is not ideal.
* When the app is killed and an Intercom notification comes in, tapping on the notification launches the app but does not show the messenger window. However, the messenger bubble appears.

**Comment by jtreanor** [comment]
Your first point is valid. I agree we should be consistent with these callbacks. It should be intercepted along with `userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:`.

The second point is interesting. Does this mean Intercom notifications are not handled correctly both when coming from a backgrounded and from a killed state? If so this is not the issue I thought it was.

**Comment by jtreanor** [comment]
Have you tried this out on iOS 9?

**Comment by jtreanor** [comment]
Apologies for the repeated messages.
I've just had an idea. Its possible that this is not a push notification issue. Perhaps the messenger just isn't correctly visible.

Does `userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:` and/or `application:didReceiveRemoteNotification:fetchCompletionHandler:` get called for Intercom push notifications?

**Comment by phatmann** [comment]
If the app is in the background or killed when I tap on an Intercom notification, then `userNotificationCenter:didReceive` is getting called. If in foreground, `userNotificationCenter:willPresent` is called. Isn't Intercom supposed to intercept these calls and not pass them on?

A suggestion: it would be better if you guys did not swizzle these methods, and instead the developers had to wire them up, exactly as you have to on Android. Less magic and more control is always better.

**Comment by jtreanor** [comment]
Yes, Intercom is supposed to intercept these. If `application:didReceiveRemoteNotification:fetchCompletionHandler:` also does get called it means that for some reason the swizzling has failed. This is exceedingly rare. Do you see this line in your logs?

```
[Intercom] ERROR - Could not find the application delegate's class. Push notifications may not work correctly.
```

Thanks for the suggestion. I hear you. We get mixed feedback on this. Many people love the simplicity of getting set up with one line while others understandably would like more control. We will explore finding a middle ground or perhaps making it optional in a future release.

**Comment by phatmann** [comment]
I do not see that error. However, my AppDelegate base class is in an embedded framework (we have two similar apps so we need to share the code), so it would not surprise me if Intercom somehow failed to swizzle it. What is the suggested workaround? Perhaps if I explicitly declared these methods in the derived class the swizzling would work? I will play around with it and let you know.

**Comment by jtreanor** [comment]
That is almost certainly the issue.

The reason for this is that prior to performing the swizzle we do this check:
```
if ((class_conformsToProtocol(class, @protocol(UIApplicationDelegate)) || class_conformsToProtocol(class, @protocol(UNUserNotificationCenterDelegate)))
    && [NSBundle bundleForClass:class] == [NSBundle mainBundle]) {
    // Logic here
}
```

The `NSBundle` check is there so that we don't interfere with any other third party component that has classes conforming to those protocols. Its quite common for push notification libraries to use app delegate proxies.
However, the valid use case you describe suggests we should reconsider this. I'll chat to the team about this on Monday and decide how to proceed. We will certainly get a solution for you ASAP.

Unfortunately, I can't think of a quick workaround here. `class_conformsToProtocol` will not be true unless the class itself declares that it implements the protocol. Redundant protocol declarations give build errors with Swift.

One thing you could do is compose the base app delegate rather than subclass it.

**Comment by phatmann** [comment]
I will just duplicate the relevant code in each app delegate. No shame in copy and paste with a purpose 😬. There is already a shared class to handle notifications so it's very little dup code.

**Comment by jtreanor** [comment]
Unfortunately there are some issues with 3.0.14 on iOS 8 and iOS 9. Because of this, we have just release [3.0.15](https://github.com/intercom/intercom-ios/releases/tag/3.0.15) which reverts the `UserNotifications` changes.
I recommend against using 3.0.14 if you need to support iOS 8/9. We will release an update that will fully close out this issue on Monday (Irish time).

**Comment by jtreanor** [comment]
Hi @phatmann 👋 

[3.0.16](https://github.com/intercom/intercom-ios/releases/tag/3.0.16) has now been released.

This includes both changes we discussed in this thread:
- Support for `UserNotifications`. i.e. Intercept Intercom push in `userNotificationCenter:didReceive` and `userNotificationCenter:willPresent`.
- Swizzle classes that are not in the main bundle - this means your shared app delegate will now work.

Thanks for all you help. Let me know if you are still seeing issues with this.
